lib/src/signing.rs

```diff
@@ -0,0 +1,217 @@
+/// versions of them.
+pub trait SigningBackend: Debug + Send + Sync {
+    /// The name of the backend, used for selecting the main backend in config
+    /// and in CLI.
+    fn name(&self) -> &str;
+
+    /// Check if the signature was created by this backend implementation.
+    ///
+    /// Should check the signature format, usually just looks at the prefix.
+    fn is_of_this_type(&self, signature: &[u8]) -> bool;
+
+    /// Create a signature for arbitrary data.
+    ///
+    /// The `key` parameter is what `jj sign` receives as key argument, or what
+    /// is configured in the `signing.key` config.
+    fn sign(&self, data: &[u8], key: Option<&str>) -> Result<Vec<u8>, SignError>;
+
+    /// Verify a signature. Should be reflexive with `sign`:
+    /// ```rust,ignore
+    /// verify(data, sign(data)?)?.status == SigStatus::Good
+    /// ```
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<Verification, SignError>;
+}
+
+#[derive(Debug, Error)]
+pub enum SignError {
+    #[error("Would override a foreign commit signature")]
+    CannotRewrite,
+    #[error("Signing error: {0}")]
+    Other(#[from] Box<dyn std::error::Error + Send + Sync>),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum SignBehavior {
+    /// Drop existing signatures.
+    /// This is what jj did before signing support or does now when a signing
+    /// backend is not configured.
+    #[default]
+    Drop,
+    /// Only sign commits that were authored by self and already signed,
+    /// "preserving" the signature across rewrites.
+    /// This is what jj does when a signing backend is configured.
+    Keep,
+    /// Sign/re-sign commits that were authored by self. Foreign signatures are
+    /// overwritten on authored commits and dropped on others.
+    /// This is what jj does when configured to always sign.
+    Own,
+    /// Always sign commits, regardless of who authored or signed them before.
+    /// This is what jj does on `jj sign -f` or `jj --force-sign <snapshotting
+    /// command>`.
+    Force,
+}
+
+impl SignBehavior {
+    pub fn from_config(sign_all: bool) -> Self {
+        match sign_all {
+            true => SignBehavior::Own,
+            false => SignBehavior::Keep,
+        }
+    }
+}
+
+/// Wraps low-level signing backends and adds caching, similar to `Store`.
+#[derive(Debug, Default)]
+pub struct Signer {
+    main_backend: Option<Box<dyn SigningBackend>>,
+    other_backends: Vec<Box<dyn SigningBackend>>,
+    cache: RwLock<HashMap<CommitId, Verification>>,
+}
+
+impl Signer {
+    pub fn from_settings(settings: &UserSettings) -> Self {
+        let mut other_backends: Vec<Box<dyn SigningBackend>> = vec![
+            // todo implement some backends
+        ];
+
+        let main_backend = settings
+            .signing_backend()
+            .and_then(|backend| other_backends.iter().position(|b| b.name() == backend))
+            .map(|pos| other_backends.swap_remove(pos));
```

Comment:
nit: might be easier to build a map of `name: backend`. We'll probably need to report an error if the configured backend name is wrong.

(Error handling can be added later.)

Reply:
Meh, with there being like 3 backends total (maybe a bit more but still a tiny number) a linear search is perfectly fine if not better.

Although with a map there's no need to have `SigningBackend::name` anymore, hmm

For now I made it into a map, but btw I feel like some out-of-git cache of verifications will have to be made at some point in the future as an optimisation to not call the backend (aka subprocess into `gpg` of `ssh-keygen` or whatever) for every single signed commit - right now if some repo had a large number of those the performance would be awful (I'm planning to add the signature checkmark to default commit log template)